File: Tests/iOS Tests/Tests/PNPublishSizeOfMessage.m
//
//  PNPublishSizeOfMessage.m
//  PubNub Tests
//
//  Created by Jordan Zucker on 3/23/16.
//
//

#import <PubNub/PubNub.h>

#import "PNBasicClientTestCase.h"

#import "NSString+PNTest.h"

@interface PNPublishSizeOfMessage : PNBasicClientTestCase
@end

@implementation PNPublishSizeOfMessage

- (BOOL)isRecording {
    return NO;
}

- (void)setUp {
    [super setUp];
    PNConfiguration *config = [PNConfiguration configurationWithPublishKey:self.publishKey
                                                              subscribeKey:self.subscribeKey];
    config.uuid = @"322A70B3-F0EA-48CD-9BB0-D3F0F5DE996C";
    self.client = [PubNub clientWithConfiguration:config];
}

- (void)tearDown {
    self.client = nil;
    [super tearDown];
}

- (void)testSizeOfMessageToChannel {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSString *message = @"test";
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 494, 20, @"Size is different than expected %@ <> %@ +-20", @(size), @(494));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeOfMessageToNilChannel {
    
    NSString *message = @"test";
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    #pragma clang diagnostic push
    #pragma clang diagnostic ignored "-Wnonnull"
    [self.client sizeOfMessage:message
                     toChannel:nil
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqual(size, -1, @"Size is different than expected %@ <> %@", @(size), @(-1));
                    
                    [completionBlockExpectation fulfill];
                }];
    #pragma clang diagnostic pop
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeOfMessageStoreInHistory {
    
    NSString *message = @"test";
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    #pragma clang diagnostic push
    #pragma clang diagnostic ignored "-Wnonnull"
    [self.client sizeOfMessage:message
                     toChannel:nil
                storeInHistory:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqual(size, -1, @"Size is different than expected %@ <> %@", @(size), @(-1));
                    
                    [completionBlockExpectation fulfill];
                }];
    #pragma clang diagnostic pop
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeOfMessageCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSString *message = @"test";
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 603, 20, @"Size is different than expected %@ <> %@ +/-20", @(size), @(603));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeOfMessageStoreInHistoryAndCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSString *message = @"test";
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:YES
                storeInHistory:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 603, 20, @"Size is different than expected %@ <> %@ +/-20", @(size), @(603));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeOfMessageStoreInHistoryNotCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSString *message = @"test";
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:NO
                storeInHistory:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 494, 20, @"Size is different than expected %@ <> %@ +/-20", @(size), @(494));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeOfMessageNotStoreInHistoryNotCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSString *message = @"test";
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:NO
                storeInHistory:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 494, 20, @"Size is different than expected %@ <> %@ +/-20", @(size), @(494));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeOfMessageNotStoreInHistoryCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSString *message = @"test";
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:YES
                storeInHistory:NO
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 611, 20, @"Size is different than expected %@ <> %@  +/-20", @(size), @(611));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSize10kMessageStoreInHistoryCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSString *message = [NSString randomAlphanumericStringWithLength:10000];

    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:YES
                storeInHistory:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 8119, 500, @"Size is different than expected %@ <> %@  +/-500", @(size), @(8119));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSize100kMessageStoreInHistoryCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSString *message = [NSString randomAlphanumericStringWithLength:100000];
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:YES
                storeInHistory:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 75793, 5000, @"Size is different than expected %@ <> %@ +/-5000", @(size), @(75793));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeDictionaryMessageStoreInHistoryCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSDictionary *message = @{@"1": @"3", @"2": @"3"};
    
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:YES
                storeInHistory:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 610, 20, @"Size is different than expected %@ <> %@ +/-20", @(size), @(610));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeNestedDictionaryMessageStoreInHistoryCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSDictionary *message = @{@"1": @{@"1": @{@"3": @"5"}}, @"2": @"3"};
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:YES
                storeInHistory:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 620, 20, @"Size is different than expected %@ <> %@ +/-20", @(size), @(620));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeArrayMessageStoreInHistoryCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSArray *message = @[@"1", @"2", @"3", @"4"];
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:YES
                storeInHistory:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 611, 20, @"Size is different than expected %@ <> %@ +/-20", @(size), @(611));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

- (void)testSizeComplexArrayMessageStoreInHistoryCompressed {
    
    NSString *channelName = [[NSUUID UUID] UUIDString];
    NSArray *message =   @[@"1", @{@"1": @{@"1": @"2"}}, @[@"1", @"2", @(2)], @(567)];
    
    XCTestExpectation *completionBlockExpectation = [self expectationWithDescription:@"Completion"];
    
    [self.client sizeOfMessage:message
                     toChannel:channelName
                    compressed:YES
                storeInHistory:YES
                withCompletion:^(NSInteger size) {
                    NSLog(@"%@", @(size));
                    
                    XCTAssertEqualWithAccuracy(size, 626, 20, @"Size is different than expected %@ <> %@ +/-20", @(size), @(626));
                    
                    [completionBlockExpectation fulfill];
                }];
    
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

@end

File: Tests/iOS Tests-Swift/PNPublishSizeOfMessage.swift
//
//  PNPublishSizeOfMessage.swift
//  PubNub Tests
//
//  Created by Vadim Osovets on 09/15/15.
//
//

import UIKit
import XCTest

extension String {
 
    static func randomStringWithLength (len : Int) -> NSString {
        let letters : NSString = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        
        let randomString : NSMutableString = NSMutableString(capacity: len)
        
        for (var i=0; i < len; i++){
            let length = UInt32 (letters.length)
            let rand = arc4random_uniform(length)
            randomString.appendFormat("%C", letters.characterAtIndex(Int(rand)))
        }
        
        return randomString
    }
}

class PNPublishSizeOfMessage: PNBasicClientTestCase {
    
    func testSizeOfMessageToChannel() {
        let channelName = NSUUID().UUIDString
        let message = "test"

        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName) { (size: Int) -> Void in
            XCTAssertTrue(abs(size - 376) <= 20)
            expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }

    func testSizeOfMessageToNilChannel() {
        let message = "test"
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: nil) { (size: Int) -> Void in
                XCTAssertEqual(size, -1, "Sizes are not equal")
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }

    func testSizeOfMessageStoreInHistory() {
        let message = "test"
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message, toChannel: nil, storeInHistory: true) { (size: Int) -> Void in
            XCTAssertEqual(size, -1, "Sizes are not equal")
            expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
    
    func testSizeOfMessageCompressed() {
        let channelName = NSUUID().UUIDString
        let message = "test"
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName, compressed: true) { (size: Int) -> Void in
                XCTAssertTrue(abs(size - 496) <= 20)
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
    
    func testSizeOfMessageStoreInHistoryAndCompressed() {
        let channelName = NSUUID().UUIDString
        let message = "test"
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName, compressed: true, storeInHistory: true) { (size: Int) -> Void in
                XCTAssertTrue(abs(size - 496) <= 20)
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
    
    func testSizeOfMessageStoreInHistoryNotCompressed() {
        let channelName = NSUUID().UUIDString
        let message = "test"
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName, compressed: false, storeInHistory: true) { (size: Int) -> Void in
                XCTAssertTrue(abs(size - 386) <= 20)
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }

    func testSizeOfMessageNotStoreInHistoryNotCompressed() {
        let channelName = NSUUID().UUIDString
        let message = "test"
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName, compressed: false, storeInHistory: false) { (size: Int) -> Void in
                XCTAssertTrue(abs(size - 386) <= 20)
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
    
    func testSizeOfMessageNotStoreInHistoryCompressed() {
        let channelName = NSUUID().UUIDString
        let message = "test"
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName, compressed: true, storeInHistory: false) { (size: Int) -> Void in
                XCTAssertTrue(abs(size - 504) <= 20)
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }

    func testSize10kMessageStoreInHistoryCompressed() {
        let channelName = NSUUID().UUIDString
        let message = String.randomStringWithLength(10000)
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName, compressed: true, storeInHistory: true) { (size: Int) -> Void in
                XCTAssertTrue(abs(size - 8000) <= 500)
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
    
    func testSize100kMessageStoreInHistoryCompressed() {
        let channelName = NSUUID().UUIDString
        let message = String.randomStringWithLength(100000)
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName, compressed: true, storeInHistory: true) { (size: Int) -> Void in
                XCTAssertTrue(abs(size - 75000) <= 5000)
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
    
    func testSizeDictionaryMessageStoreInHistoryCompressed() {
        let channelName = NSUUID().UUIDString
        let message = ["1": "3", "2": "3"]
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName, compressed: true, storeInHistory: true) { (size: Int) -> Void in
                XCTAssertTrue(abs(size - 503) <= 20)
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
    
    func testSizeNestedDictionaryMessageStoreInHistoryCompressed() {
        let channelName = NSUUID().UUIDString
        let message = ["1": ["1": ["3": "5"]], "2": "3"]
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName, compressed: true, storeInHistory: true) { (size: Int) -> Void in
                XCTAssertTrue(abs(size - 513) <= 20)
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }

    
    func testSizeArrayMessageStoreInHistoryCompressed() {
        let channelName = NSUUID().UUIDString
        let message = ["1", "2", "3", "4"]
        
        let expectation = self.expectationWithDescription("Completion")
        
        self.client.sizeOfMessage(message,
            toChannel: channelName, compressed: true, storeInHistory: true) { (size: Int) -> Void in
                XCTAssertTrue(abs(size - 504) <= 20)
                expectation.fulfill()
        }
        
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
}
